File: modules/gpu/src/metal/buffer.m
#include <string.h>

#include "metal.h"

Mel_Gpu_Buffer* mel_gpu_buffer_create_opt(Mel_Gpu_Device* dev, Mel_Gpu_Buffer_Opt opt)
{
    if (!dev || opt.size == 0) return NULL;

    Mel_Gpu_Buffer* buf = calloc(1, sizeof *buf);
    if (!buf) return NULL;
    buf->size         = opt.size;
    buf->host_visible = (opt.memory != MEL_GPU_MEMORY_GPU_ONLY);

    MTLResourceOptions res = buf->host_visible ? MTLResourceStorageModeShared
                                               : MTLResourceStorageModePrivate;

    if (buf->host_visible) {
        buf->mtl = opt.data ? [dev->mtl newBufferWithBytes:opt.data length:opt.size options:res]
                            : [dev->mtl newBufferWithLength:opt.size options:res];
    } else {
        buf->mtl = [dev->mtl newBufferWithLength:opt.size options:res];
        if (buf->mtl && opt.data) {
            id<MTLBuffer> staging = [dev->mtl newBufferWithBytes:opt.data length:opt.size
                                                         options:MTLResourceStorageModeShared];
            id<MTLCommandBuffer>      cb   = [dev->queue commandBuffer];
            id<MTLBlitCommandEncoder> blit = [cb blitCommandEncoder];
            [blit copyFromBuffer:staging sourceOffset:0 toBuffer:buf->mtl destinationOffset:0 size:opt.size];
            [blit endEncoding];
            [cb commit];
            [cb waitUntilCompleted];
        }
    }

    if (!buf->mtl) { free(buf); return NULL; }
    return buf;
}

void mel_gpu_buffer_destroy(Mel_Gpu_Buffer* buf)
{
    if (!buf) return;
    buf->mtl = nil;
    free(buf);
}

void* mel_gpu_buffer_map(Mel_Gpu_Buffer* buf)
{
    if (!buf || !buf->host_visible) return NULL;
    return [buf->mtl contents];
}

void mel_gpu_buffer_write(Mel_Gpu_Buffer* buf, const void* data, usize size)
{
    if (!buf || !buf->host_visible || !data || size == 0) return;
    if (size > buf->size) size = buf->size;
    memcpy([buf->mtl contents], data, size);
}

File: modules/gpu/src/webgpu/buffer.c
#include "webgpu_backend.h"

Mel_Gpu_Buffer* mel_gpu_buffer_create_opt(Mel_Gpu_Device* dev, Mel_Gpu_Buffer_Opt opt)
{
    if (!dev || opt.size == 0) return NULL;

    WGPUBufferUsage usage = WGPUBufferUsage_CopyDst;
    if (opt.usage & MEL_GPU_BUFFER_VERTEX)  usage |= WGPUBufferUsage_Vertex;
    if (opt.usage & MEL_GPU_BUFFER_INDEX)   usage |= WGPUBufferUsage_Index;
    if (opt.usage & MEL_GPU_BUFFER_UNIFORM) usage |= WGPUBufferUsage_Uniform;

    usize size = (opt.size + 3u) & ~(usize)3u; // WebGPU buffer sizes are 4-byte aligned

    WGPUBufferDescriptor d = {
        .usage            = usage,
        .size             = size,
        .mappedAtCreation = opt.data != NULL,
    };
    WGPUBuffer b = wgpuDeviceCreateBuffer(dev->device, &d);
    if (!b) return NULL;

    if (opt.data) {
        void* p = wgpuBufferGetMappedRange(b, 0, size);
        if (p) memcpy(p, opt.data, opt.size);
        wgpuBufferUnmap(b);
    }

    Mel_Gpu_Buffer* buf = calloc(1, sizeof *buf);
    if (!buf) { wgpuBufferRelease(b); return NULL; }
    buf->device       = dev;
    buf->buf          = b;
    buf->size         = size;
    buf->host_visible = (opt.memory != MEL_GPU_MEMORY_GPU_ONLY);
    return buf;
}

void mel_gpu_buffer_destroy(Mel_Gpu_Buffer* buf)
{
    if (!buf) return;
    if (buf->buf) wgpuBufferRelease(buf->buf);
    free(buf);
}

void* mel_gpu_buffer_map(Mel_Gpu_Buffer* buf)
{
    (void)buf;
    return NULL; // WebGPU mapping is asynchronous; use mel_gpu_buffer_write
}

void mel_gpu_buffer_write(Mel_Gpu_Buffer* buf, const void* data, usize size)
{
    if (!buf || !data || size == 0) return;
    if (size > buf->size) size = buf->size;
    size &= ~(usize)3u; // wgpuQueueWriteBuffer requires a 4-byte-multiple size
    if (size == 0) return;
    wgpuQueueWriteBuffer(buf->device->queue, buf->buf, 0, data, size);
}
